cli/tools/test.rs

```diff
@@ -967,6 +1079,164 @@ impl TestReporter for PrettyTestReporter {
+    &self,
+    elapsed: &Duration,
+    tests: &IndexMap<usize, TestDescription>,
+  ) {
+    let mut suites: IndexMap<String, quick_junit::TestSuite> = IndexMap::new();
+    for (id, case) in &self.cases {
+      if let Some(test) = tests.get(id) {
+        suites
+          .entry(test.location.file_name.clone())
+          .and_modify(|s| {
+            s.add_test_case(case.clone());
+          })
+          .or_insert_with(|| {
+            quick_junit::TestSuite::new(test.location.file_name.clone())
+              .add_test_case(case.clone())
+              .to_owned()
+          });
+      }
+    }
+
+    let mut report = quick_junit::Report::new("deno test");
+    report.set_time(*elapsed).add_test_suites(
+      suites
+        .values()
+        .cloned()
+        .collect::<Vec<quick_junit::TestSuite>>(),
+    );
+
+    // How should we handle errors when writing out the report if at all?
+    if let Ok(mut file) = std::fs::File::create(self.path.clone()) {
+      let _ = file.write_all(report.to_string().unwrap().as_bytes());
+    }
+  }
+}
+
+impl TestReporter for JunitTestReporter {
+  fn report_register(&mut self, description: &TestDescription) {
+    self.cases.insert(
+      description.id,
+      quick_junit::TestCase::new(
+        description.name.clone(),
+        quick_junit::TestCaseStatus::skipped(),
+      ),
+    );
+  }
+
+  fn report_plan(&mut self, _plan: &TestPlan) {}
+
+  fn report_wait(&mut self, _description: &TestDescription) {}
+
+  fn report_output(&mut self, _output: &[u8]) {
+    /*
+     TODO: it would be nice to include stdout/stderr, but we currently have a
+     global stream for both that doesn't keep track of what test produced what.
+    */
+  }
+
+  fn report_result(
+    &mut self,
+    description: &TestDescription,
+    result: &TestResult,
+    elapsed: u64,
+  ) {
+    if let Some(mut case) = self.cases.get_mut(&description.id) {
+      case.status = Self::convert_status(result);
+      case.set_time(Duration::from_millis(elapsed));
+    }
+  }
+
+  fn report_uncaught_error(&mut self, _origin: &str, _error: Box<JsError>) {}
+
+  fn report_step_register(&mut self, _description: &TestStepDescription) {}
+
+  fn report_step_wait(&mut self, _description: &TestStepDescription) {}
+
+  fn report_step_result(
+    &mut self,
+    description: &TestStepDescription,
+    result: &TestStepResult,
+    _elapsed: u64,
+    _tests: &IndexMap<usize, TestDescription>,
+    _test_steps: &IndexMap<usize, TestStepDescription>,
+  ) {
+    let status = match result {
+      TestStepResult::Ok => "passed",
+      TestStepResult::Ignored => "skipped",
+      TestStepResult::Failed(_) => "failure",
+    };
+    // TODO: This currently doesn't work for nested steps, not sure how I'd represent them...
+    if let Some(case) = self.cases.get_mut(&description.parent_id) {
+      case.add_property(quick_junit::Property::new(
+        format!("step[{}]", status),
+        description.name.clone(),
+      ));
```

Comment:
Do we care about representing nested steps?

Reply:
So the JUnit hierarchy looks something like:
- `testsuites`, main container that holds everything else
- `testsuite`, currently I'm treating files as test suites since we don't have a container above `Deno.test`
- `testcases`, the actual core unit can pass/fail/skip. Currently every `Deno.test` call is it's own test, which I think is the correct method

Steps aren't actually a core type but are implemented in frameworks that have them by adding properties to the test case with key `step[pass|failure|skipped]` and the name of the step as the value. Properties are a flat structure though, so there's no native way to represent the nesting.

The least bad thing I can think of is to do something like this
```xml
 <testcase name="parentOnTextContext" time="0.028">
    <properties>
        <property name="step[passed]" value="step"/>
        <property name="step[passed]" value="step ... nested step"/>
    </properties>
</testcase>
```

Reply:
Maybe we could make test cases into `testcase` where the name is a concatenation of the parent test case name and the test step name?

Reply:
I think the core reason I'm having such a hard time building a mental model for this that works is that Deno's test setup allows for nesting of tests themselves (at this point I feel like steps are just nestable tests). I've only ever seen nesting of test suites (i.e. collections of tests, but not the tests themselves), hence my uncertainty in how to map this.

Here's a toy example
```js
# example_test.ts
Deno.test('simple test', () => {
  assertEquals(1 + 2, 3);
}

Deno.test('test with steps', (t) => {
  t.step('nested', (t) => {
      assertEquals(2 + 3, 5);
  }
  assertEquals(3 + 4, 7);
}
```

The most naive way I can think of representing where steps are treated as a `testcase` is the following, but it has an issue of double counting successes and failures (i.e. if `nested` fails, `test with steps` will also fail and we'll count 2 failures)
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="deno test" tests="7" failures="0" errors="0" time="0.170">
    <testsuite name="example_test.ts" tests="3" disabled="0" errors="0" failures="0">
        <testcase name="simple test"></testcase>
        <testcase name="test with steps"></testcase>
        <testcase name="test with steps ... nested">
    </testsuite>
</testsuites>
```

Reply:
Decided to implement the first approach in [7cf41ff](https://github.com/denoland/deno/pull/19747/commits/7cf41ff92727a5ee2e3d0c0de9d83b10a7d4df4e). 

It's not the ideal approach, but it's the least bad that I can think of at the moment.
```xml
<testcase name="parentOnTextContext" time="0.030">
    <properties>
        <property name="step[passed]" value="step ... nested step"/>
        <property name="step[passed]" value="step"/>
    </properties>
</testcase>
```